pkg/virt-controller/watch/clone/clone.go

```diff
@@ -0,0 +1,334 @@
+	failEvent      Event
+	failReason     string
+
+	reenqueueInfo reenqueueInfo
+}
+type reenqueueInfo struct {
+	reenqueueReason string
+}
+
+func (ctrl *VMCloneController) execute(key string) (error, reenqueueInfo) {
+	var syncInfo syncInfo
+	doNotReenqueue := reenqueueInfo{}
+	logger := log.Log
+
+	obj, cloneExists, err := ctrl.vmCloneInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return err, doNotReenqueue
+	}
+
+	var vmClone *clonev1alpha1.VirtualMachineClone
+	if cloneExists {
+		vmClone = obj.(*clonev1alpha1.VirtualMachineClone)
+		logger = logger.Object(vmClone)
+	} else {
+		return nil, doNotReenqueue
+	}
+
+	var syncErr error
+
+	sourceInfo := vmClone.Spec.Source
+	switch cloneSourceType(sourceInfo.Kind) {
+	case sourceTypeVM:
+		vmKey := getKey(sourceInfo.Name, vmClone.Namespace)
+		obj, vmExists, err := ctrl.vmInformer.GetStore().GetByKey(vmKey)
+		if err != nil {
+			return fmt.Errorf("error getting VM %s in namespace %s from cache: %v", sourceInfo.Name, vmClone.Namespace, err), doNotReenqueue
+		}
+		if !vmExists {
+			return fmt.Errorf("VM %s in namespace %s does not exist", sourceInfo.Name, vmClone.Namespace), doNotReenqueue
+		}
+		sourceVM := obj.(*k6sv1.VirtualMachine)
+
+		syncInfo = ctrl.syncSourceVM(key, sourceVM, vmClone)
+		syncErr = syncInfo.err
+	default:
+		return fmt.Errorf("clone %s is defined with an unknown source type %s", vmClone.Name, sourceInfo.Kind), doNotReenqueue
+	}
+
+	err = ctrl.updateStatus(vmClone, syncInfo)
+	if err != nil {
+		return fmt.Errorf("error updating status: %v", err), syncInfo.reenqueueInfo
+	}
+
+	if syncErr != nil {
+		return fmt.Errorf("sync error: %v", syncErr), syncInfo.reenqueueInfo
+	}
+
+	return nil, syncInfo.reenqueueInfo
+}
+
+func (ctrl *VMCloneController) syncSourceVM(key string, source *k6sv1.VirtualMachine, vmClone *clonev1alpha1.VirtualMachineClone) syncInfo {
+	var targetType cloneTargetType
+	if vmClone.Spec.Target != nil {
+		targetType = cloneTargetType(vmClone.Spec.Target.Kind)
+	} else {
+		targetType = defaultType
+	}
+
+	switch targetType {
+	case targetTypeVM:
+		return ctrl.syncSourceVMTargetVM(key, source, vmClone)
+
+	default:
+		return syncInfo{err: fmt.Errorf("target type is unknown: %s", targetType)}
+	}
+}
+
+func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6sv1.VirtualMachine, vmClone *clonev1alpha1.VirtualMachineClone) syncInfo {
+	syncInfo := syncInfo{}
+	logger := log.Log.Object(vmClone)
+
+	var snapshot *snapshotv1alpha1.VirtualMachineSnapshot
+	targetVMInfo := vmClone.Spec.Target
+
+	switch vmClone.Status.Phase {
+	case clonev1alpha1.PhaseUnset, clonev1alpha1.SnapshotInProgress:
+
+		// Create snapshot
+		if vmClone.Status.SnapshotName == nil {
+			snapshot = generateSnapshot(vmClone, source)
+			logger.Infof("creating snapshot %s for clone %s", snapshot.Name, vmClone.Name)
+
+			snapshot, syncInfo.err = ctrl.client.VirtualMachineSnapshot(snapshot.Namespace).Create(context.Background(), snapshot, v1.CreateOptions{})
```

Comment:
This snapshot does not appear to get deleted.  At a minimum, it should be deleted when the VMClone resource is deleted via ownerref).  But I think it would be better for it to be deleted once the restore is complete.

Reply:
Sounds right. Thanks a lot. :+1:

Reply:
Done :ok_hand: